### ISSUE TITLE
"spack reindex" fails with "invalid ref count"

### ISSUE THREAD

**Issue by eschnett** [LLNL/spack#1720]
A previous install of `git` failed, leaving a half-installed `git` ghost. This is usually remedied via `spack reindex`, but this fails here.

My original command is:
```
$ spack install -v julia
==> Installing julia
==> git is already installed in /home/eschnett/src/spack/opt/spack/linux-centos7-x86_64/gcc-6.1.0-spack/git-2.9.3-kgah4b5gurqx2rswlt44xoinshfw7doi
Traceback (most recent call last):
  File "/home/eschnett/src/spack/bin/spack", line 192, in <module>
    main()
  File "/home/eschnett/src/spack/bin/spack", line 169, in main
    return_val = command(parser, args)
  File "/home/eschnett/src/spack/lib/spack/spack/cmd/install.py", line 92, in install
    explicit=True)
  File "/home/eschnett/src/spack/lib/spack/spack/package.py", line 919, in do_install
    run_tests=run_tests)
  File "/home/eschnett/src/spack/lib/spack/spack/package.py", line 1069, in do_install_dependencies
    dep.package.do_install(**kwargs)
  File "/home/eschnett/src/spack/lib/spack/spack/package.py", line 901, in do_install
    rec = spack.installed_db.get_record(self.spec)
  File "/home/eschnett/src/spack/lib/spack/spack/database.py", line 75, in converter
    return function(self, spec_like, *args, **kwargs)
  File "/home/eschnett/src/spack/lib/spack/spack/database.py", line 521, in get_record
    key = self._get_matching_spec_key(spec, **kwargs)
  File "/home/eschnett/src/spack/lib/spack/spack/database.py", line 516, in _get_matching_spec_key
    raise KeyError("No such spec in database! %s" % spec)
KeyError: 'No such spec in database! git@2.9.3%gcc@6.1.0-spack arch=linux-centos7-x86_64^autoconf@2.69%gcc@6.1.0-spack arch=linux-centos7-x86_64^bzip2@1.0.6%gcc@6.1.0-spack arch=linux-centos7-x86_64^curl@7.50.1%gcc@6.1.0-spack arch=linux-centos7-x86_64^expat@2.2.0%gcc@6.1.0-spack arch=linux-centos7-x86_64^gettext@0.19.8.1%gcc@6.1.0-spack+bzip2+curses+git~libunistring+libxml2+tar+xz arch=linux-centos7-x86_64^libsigsegv@2.10%gcc@6.1.0-spack arch=linux-centos7-x86_64^libxml2@2.9.4%gcc@6.1.0-spack~python arch=linux-centos7-x86_64^m4@1.4.17%gcc@6.1.0-spack+sigsegv arch=linux-centos7-x86_64^ncurses@6.0%gcc@6.1.0-spack arch=linux-centos7-x86_64^openssl@1.0.2h%gcc@6.1.0-spack arch=linux-centos7-x86_64^pcre@8.39%gcc@6.1.0-spack+utf arch=linux-centos7-x86_64^perl@5.24.0%gcc@6.1.0-spack+cpanm arch=linux-centos7-x86_64^tar@1.29%gcc@6.1.0-spack arch=linux-centos7-x86_64^xz@5.2.2%gcc@6.1.0-spack arch=linux-centos7-x86_64^zlib@1.2.8%gcc@6.1.0-spack arch=linux-centos7-x86_64'
```

To remedy, I try to reindex:
```
$ spack reindex
Traceback (most recent call last):
  File "/home/eschnett/src/spack/bin/spack", line 192, in <module>
    main()
  File "/home/eschnett/src/spack/bin/spack", line 169, in main
    return_val = command(parser, args)
  File "/home/eschnett/src/spack/lib/spack/spack/cmd/reindex.py", line 31, in reindex
    spack.installed_db.reindex(spack.install_layout)
  File "/home/eschnett/src/spack/lib/spack/spack/database.py", line 375, in reindex
    self._check_ref_counts()
  File "/home/eschnett/src/spack/lib/spack/spack/database.py", line 407, in _check_ref_counts
    (key, found, expected, self._index_path))
AssertionError: Invalid ref_count: 3f2b2nmz2brdouohnoxiik3zch7d4v7p: 31 (expected 30), in DB /home/eschnett/src/spack/opt/spack/.spack-db/index.yaml
```

**Comment by tgamblin** [comment]
@eschnett: I can't reproduce this one.  Can you pull the latest `develop` and attach the results of `spack debug create-db-tarball` when this happens again?

**Comment by eschnett** [comment]
This error happens reliably. I attached the tarball. The two previous (failing) commands were
```
$ spack reindex                                                                 Traceback (most recent call last):
  File "/home/eschnett/src/spack/bin/spack", line 192, in <module>
    main()
  File "/home/eschnett/src/spack/bin/spack", line 169, in main
    return_val = command(parser, args)
  File "/home/eschnett/src/spack/lib/spack/spack/cmd/reindex.py", line 31, in reindex
    spack.installed_db.reindex(spack.install_layout)
  File "/home/eschnett/src/spack/lib/spack/spack/database.py", line 375, in reindex
    self._check_ref_counts()
  File "/home/eschnett/src/spack/lib/spack/spack/database.py", line 407, in _check_ref_counts
    (key, found, expected, self._index_path))
AssertionError: Invalid ref_count: 3f2b2nmz2brdouohnoxiik3zch7d4v7p: 31 (expected 30), in DB /home/eschnett/src/spack/opt/spack/.spack-db/index.yaml
eschnett@bethe:~/src/spack (08:37:11)
$ spack install -v julia                                                        ==> Installing julia
==> git is already installed in /home/eschnett/src/spack/opt/spack/linux-centos7-x86_64/gcc-6.1.0-spack/git-2.9.3-kgah4b5gurqx2rswlt44xoinshfw7doi
Traceback (most recent call last):
  File "/home/eschnett/src/spack/bin/spack", line 192, in <module>
    main()
  File "/home/eschnett/src/spack/bin/spack", line 169, in main
    return_val = command(parser, args)
  File "/home/eschnett/src/spack/lib/spack/spack/cmd/install.py", line 92, in install
    explicit=True)
  File "/home/eschnett/src/spack/lib/spack/spack/package.py", line 919, in do_install
    run_tests=run_tests)
  File "/home/eschnett/src/spack/lib/spack/spack/package.py", line 1069, in do_install_dependencies
    dep.package.do_install(**kwargs)
  File "/home/eschnett/src/spack/lib/spack/spack/package.py", line 901, in do_install
    rec = spack.installed_db.get_record(self.spec)
  File "/home/eschnett/src/spack/lib/spack/spack/database.py", line 75, in converter
    return function(self, spec_like, *args, **kwargs)
  File "/home/eschnett/src/spack/lib/spack/spack/database.py", line 521, in get_record
    key = self._get_matching_spec_key(spec, **kwargs)
  File "/home/eschnett/src/spack/lib/spack/spack/database.py", line 516, in _get_matching_spec_key
    raise KeyError("No such spec in database! %s" % spec)
KeyError: 'No such spec in database! git@2.9.3%gcc@6.1.0-spack arch=linux-centos7-x86_64^autoconf@2.69%gcc@6.1.0-spack arch=linux-centos7-x86_64^bzip2@1.0.6%gcc@6.1.0-spack arch=linux-centos7-x86_64^curl@7.50.1%gcc@6.1.0-spack arch=linux-centos7-x86_64^expat@2.2.0%gcc@6.1.0-spack arch=linux-centos7-x86_64^gettext@0.19.8.1%gcc@6.1.0-spack+bzip2+curses+git~libunistring+libxml2+tar+xz arch=linux-centos7-x86_64^libsigsegv@2.10%gcc@6.1.0-spack arch=linux-centos7-x86_64^libxml2@2.9.4%gcc@6.1.0-spack~python arch=linux-centos7-x86_64^m4@1.4.17%gcc@6.1.0-spack+sigsegv arch=linux-centos7-x86_64^ncurses@6.0%gcc@6.1.0-spack arch=linux-centos7-x86_64^openssl@1.0.2h%gcc@6.1.0-spack arch=linux-centos7-x86_64^pcre@8.39%gcc@6.1.0-spack+utf arch=linux-centos7-x86_64^perl@5.24.0%gcc@6.1.0-spack+cpanm arch=linux-centos7-x86_64^tar@1.29%gcc@6.1.0-spack arch=linux-centos7-x86_64^xz@5.2.2%gcc@6.1.0-spack arch=linux-centos7-x86_64^zlib@1.2.8%gcc@6.1.0-spack arch=linux-centos7-x86_64'
```

[spack-db.develop.f34dd94.2016-09-06-084822.tar.gz](https://github.com/LLNL/spack/files/457125/spack-db.develop.f34dd94.2016-09-06-084822.tar.gz)

**Comment by tgamblin** [comment]
Do these four directories have successful git installs in them?  They appear to have `spec.yaml` files, but I'm curious whether anything else was installed.

```
opt/spack/linux-centos7-x86_64/gcc-6.1.0-spack/git-2.8.1-jkzywad7z7nagmgmaqooa3ae4fmjubns/
opt/spack/linux-centos7-x86_64/gcc-6.1.0-spack/git-2.8.1-nqkxuakqdb5odnoksi6yvolvvkk7iowd/
opt/spack/linux-centos7-x86_64/gcc-6.1.0-spack/git-2.8.1-v6rx5mebb32iuky57xzyi6degit26fcm/
opt/spack/linux-centos7-x86_64/gcc-6.1.0-spack/git-2.9.3-kgah4b5gurqx2rswlt44xoinshfw7doi/
```

If they *do* have partial installs, are you ever using `spack install --keep-prefix`?

**Comment by TheQueasle** [comment]
I'm running into this issue as well, but not after a failed install that I'm aware of.

I've never used `spack install --keep-prefix`

```
Traceback (most recent call last):
  File "/cluster/rcss-spack/bin/spack", line 192, in <module>
    main()
  File "/cluster/rcss-spack/bin/spack", line 169, in main
    return_val = command(parser, args)
  File "/cluster/rcss-spack/lib/spack/spack/cmd/reindex.py", line 31, in reindex
    spack.installed_db.reindex(spack.install_layout)
  File "/cluster/rcss-spack/lib/spack/spack/database.py", line 375, in reindex
    self._check_ref_counts()
  File "/cluster/rcss-spack/lib/spack/spack/database.py", line 407, in _check_ref_counts
    (key, found, expected, self._index_path))
AssertionError: Invalid ref_count: x3atd7mkaebyjjs52sfhhtjrfy5l2pnz: 1 (expected 0), in DB /cluster/rcss-spack/opt/spack/.spack-db/index.yaml
```
If it helps, I can attach a db tarbal as well.

**Comment by eschnett** [comment]
@tgamblin Yes, these four `git` directories seem to contain complete installations.

No, I haven't used `--keep-prefix` with git recently, and not on this system. I don't think these are partial installs. (If that is an issue, then Spack could leave a big human-readable marker in the directory, marking it as incomplete?)

**Comment by citibeth** [comment]
Is this fixed by #1688?

**Comment by eschnett** [comment]
It is not fixed. The current head of the development branch still has this issue:
```
$ spack reindex
Traceback (most recent call last):
  File "/home/eschnett/src/spack/bin/spack", line 192, in <module>
    main()
  File "/home/eschnett/src/spack/bin/spack", line 169, in main
    return_val = command(parser, args)
  File "/home/eschnett/src/spack/lib/spack/spack/cmd/reindex.py", line 31, in reindex
    spack.installed_db.reindex(spack.install_layout)
  File "/home/eschnett/src/spack/lib/spack/spack/database.py", line 375, in reindex
    self._check_ref_counts()
  File "/home/eschnett/src/spack/lib/spack/spack/database.py", line 407, in _check_ref_counts
    (key, found, expected, self._index_path))
AssertionError: Invalid ref_count: 3f2b2nmz2brdouohnoxiik3zch7d4v7p: 31 (expected 30), in DB /home/eschnett/src/spack/opt/spack/.spack-db/index.yaml
```

**Comment by citibeth** [comment]
I think we need a procedure to rebuild this database from the installed packages in `opt/spack`.  Unless we use an ACID database (we don't), there will always be ways to break it.

**Comment by eschnett** [comment]
... isn't that what `reindex` does?

**Comment by citibeth** [comment]
@tgamblin I'm not sure.  But clearly, `reindex` shouldn't be confused by
some erroneous refcount if it's supposed to re-build those refcounts.

**Comment by robertdfrench** [comment]
I am also running into this. @tgamblin would another db tarball be helpful?

**Comment by citibeth** [comment]
I think this should be a high-priority bug to fix: it messes up the entire repo, and we have no workaround; once a Spack installation gets into this state, we don't know how to get it out.  That prevents uninstalling packages (even after they've been uninstalled), and really breaks everything.  We need, first and foremost, a way to get Spack out of this state and back into "clean" working order.

**Comment by adamhough** [comment]
I am getting this same issue as well.

**Comment by robertdfrench** [comment]
Other folks with the invalid refcount bug: is the hash in your error message referencing something in your packages.yaml file? 

I somehow ended up with two different hashes that both referenced my SLES-provided ncurses, because I was toying around with packages.yaml and just decided to throw something basic in there. 

I was able to salvage this spack install by changing all of the `opt/spack/linux-SuSE11-x86_64/gcc-5.3.0/*/.spack/spec.yaml` files to refer to ncurses with the same hash. 

Unfortunately, I do not know how I got in that state to begin with.
